Add routine to claim discount during cancellation

jboland88 opened the pull request:
#### Proposed Changes

* This PR builds on the offer step PR to add the logic to claim the discount presented.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* If you have tested this functionality before, you may need to follow some additional steps first: pbNhbs-2Ep-p2#comment-9629
* You will need a Jetpack site to test with and a Jetpack product/ plan purchase. Try testing with a new purchase.
* Navigate to `/me/purchases` to view the list of your purchases, select your purchase from the list.
* From the individual purchase page, click to cancel/ remove your purchase to start the cancellation flow. Example:
![Screen Shot 2022-07-28 at 9 34 41 AM](https://user-images.githubusercontent.com/18016357/181520644-d00dbcde-0e58-4268-be87-d861b2541ec2.png)

* Follow the steps in the cancellation flow, but do not remove the purchase. 
* You should see a screen similar to this one that presents a special offer:
![Screen Shot 2022-07-07 at 12 22 28 PM](https://user-images.githubusercontent.com/18016357/177872515-2eb89fce-4f49-4831-9414-178ef5cbdc0a.png)

* Make sure all the links provided in the text work as expected and that the copy makes sense.
* Take note of the discounted rate, as you'll need to compare the rate presented here to the rate displayed on your purchase after claiming the offer.
* From this screen, click on the "Get Discount" button
* The "Get Discount" button should switch to show "Getting Discount" and display a waiting/ loading animation
* If successful, you should see a confirmation screen that looks like this:
![Screen Shot 2022-07-26 at 12 05 12 PM](https://user-images.githubusercontent.com/18016357/181520580-26be7f09-42ab-40a8-8dbc-1f0ecdf4b74d.png)

* Clicking on "Back to my website" or clicking out of the modal should return you to the `/me/purchases` page.
* Locate the product purchase for your test site and click on it to view the individual purchase page.
* Confirm that the new discounted rate shows on the purchase page and matches the rate shown when the discount was claimed. You should also see information related to the offer in the "Price" section of the page:
![Screen Shot 2022-07-28 at 9 42 52 AM](https://user-images.githubusercontent.com/18016357/181520860-defd4794-5a69-4cdd-a48d-c5ae3f8de90b.png)


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have you sent any new strings for translation (PCYsg-1vr-p2) ASAP?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

jboland88 committed:
Intial data structure for claiming offer

jboland88 committed:
Merge branch 'add/offer-step-in-cancellation-flow' into add/routine-to-claim-cancellation-offer

jboland88 committed:
Add offer accepted step

jboland88 committed:
Merge branch 'add/offer-step-in-cancellation-flow' into add/routine-to-claim-cancellation-offer

jboland88 committed:
Merge branch 'add/offer-step-in-cancellation-flow' into add/routine-to-claim-cancellation-offer

jboland88 committed:
Update offer apply endpoint

jboland88 committed:
Merge branch 'add/offer-step-in-cancellation-flow' into add/routine-to-claim-cancellation-offer

monsieur-z commented:
@jboland88, I made sure to delete the user attribute mentioned in the linked comment, and created a new JN site with a VideoPress subscription. But I'm not seeing the discount screen. Do I need to sandbox the API and apply a specific patch when creating a testing site?

monsieur-z commented on client/components/marketing-survey/cancel-jetpack-form/index.tsx:
```diff
@@ -193,11 +206,16 @@ const CancelJetpackForm: React.FC< Props > = ( {
+
+		// When an offer has been accepted, redirect back to the purchases page.
+		if ( cancellationOfferApplySuccess || steps.OFFER_ACCEPTED_STEP === cancellationStep ) {
+			page.redirect( props.purchaseListUrl );
```
Should we make sure the URL is not external, here?

jboland88 replied:
Couldn't hurt to add that check, thanks!

monsieur-z commented on client/components/marketing-survey/cancel-jetpack-form/index.tsx:
```diff
@@ -286,17 +319,28 @@ const CancelJetpackForm: React.FC< Props > = ( {
+			action: 'close',
+			isPrimary: true,
+			disabled: disabled,
+			label: translate( 'Back to my website' ),
```
I find the label a bit misleading here. I would imagine I would be redirected to WPadmin instead of the purchases page.

jboland88 replied:
hmm, good point. Perhaps "back to my purchases" would be better.

monsieur-z commented on client/components/marketing-survey/cancel-jetpack-form/jetpack-cancellation-offer-accepted.tsx:
```diff
@@ -0,0 +1,38 @@
+			<FormattedHeader
+				headerText={ translate( 'Thanks for sticking with Jetpack!' ) }
+				subHeaderText={ translate(
+					'We’re happy you’ve chosen Jetpack to level-up your site. Your %(percentDiscount)d%% discount for %(productName)s will be applied next time you are billed.',
```
`%(percentDiscount)d%%` might be difficult for translators to read into. Maybe it'd be safer to inject the percentage amount with the symbol in the string. Or at least add a comment for translators.

jboland88 replied:
Good thought there 👍

monsieur-z commented on client/components/marketing-survey/cancel-jetpack-form/jetpack-cancellation-offer.tsx:
```diff
@@ -117,6 +128,37 @@ const JetpackCancellationOffer: React.FC< Props > = ( props ) => {
+			switch ( offerApplyError.code ) {
+				case 'invalid_offer':
+					return (
+						<p className="jetpack-cancellation-offer__error-text">
```
Nitpicking here: we could avoid duplicating the wrapping `p`.

monsieur-z commented on client/components/marketing-survey/cancel-jetpack-form/jetpack-cancellation-offer.tsx:
```diff
@@ -117,6 +128,37 @@ const JetpackCancellationOffer: React.FC< Props > = ( props ) => {
+			dispatch( applyCancellationOffer( siteId, purchase.id ) );
+			onGetDiscount(); // Takes care of analytics.
+		}
+	}, [ isApplyingOffer, offerApplySuccess, offerApplyError, onGetDiscount ] );
```
There are some missing dependencies here.

jboland88 replied:
good catch

monsieur-z commented on client/components/marketing-survey/cancel-jetpack-form/style.scss:
```diff
@@ -148,4 +152,9 @@ input[type='text'].form-text-input.jetpack-cancellation-survey__choice-item-text
+
+	&__error-text {
+		color: var( --studio-red-0 );
+		font-weight: bold;
```
We usually use numbers for `font-weight` :)

jboland88 replied:
nice, thanks!

monsieur-z commented on client/state/cancellation-offers/selectors/is-applying-cancellation-offer.js:
```diff
@@ -0,0 +1,6 @@
+/**
+ * Is there a pending request for applying a cancellation offer?
```
They're pretty self-explaining, but parameters are missing a description. I'm mentioning it since you've added a params description to other selectors.

monsieur-z commented:
> @jboland88, I made sure to delete the user attribute mentioned in the linked comment, and created a new JN site with a VideoPress subscription. But I'm not seeing the discount screen. Do I need to sandbox the API and apply a specific patch when creating a testing site?

For the record, everything works as expected when running the branch locally. It seems that this feature is not enabled yet when running the branch from the Calypso live environment.

monsieur-z reviewed (approved):
Everything tested as expected! I left minor comments in the code, but nothing blocking.

Here's some feedback about the UI, but no blockers either.

**Discount screen:**
- Heading could benefit from `preventWidows`, and maybe from having a `max-width`.
- Could we maybe make the new discount and prices bold, so that they stand out more?
- The links `how your subscription works` and `how to cancel` open a new modal, which forces the user to restart the cancellation process from scratch. Could we link to the associated articles directly instead?
 
<img width="1027" alt="Screen Shot 2022-08-02 at 11 38 06 AM" src="https://user-images.githubusercontent.com/1620183/182417709-cee0d27e-0fc0-43cb-a861-e24c6ce2fffb.png">

**Purchase screen:**
- This may just be my brain: it took me a little while to understand that the highlighted sentence below was actually  2 separate pieces of information. Could we reformulate it, or at least separate it in 2 lines?

<img width="226" alt="Screen Shot 2022-08-02 at 11 42 11 AM" src="https://user-images.githubusercontent.com/1620183/182418475-422a2a4b-9ffe-4b07-b9f4-0b38060d8d0d.png">

sirbrillig commented on client/components/jetpack-decorative-card/index.tsx:
```diff
@@ -0,0 +1,32 @@
+import * as React from 'react';
```
Minor, but since you're importing this only for the types, you can change this to the following to save a bunch of unused imports.

```suggestion
import type { FC } from 'react';
```

jboland88 replied:
Ah, nice one! Thanks!

sirbrillig commented on client/components/jetpack-decorative-card/index.tsx:
```diff
@@ -0,0 +1,32 @@
+		<div className="jetpack-decorative-card">
+			<div className="jetpack-decorative-card__header">
+				<img
+					alt="" // This is a decorative image
```
🤔 I'm not sure that a comment inside a JSX tag [is safe](https://stackoverflow.com/questions/30766441/how-to-use-comments-in-react). I think you have to escape it to avoid it ending up in the HTML.

```suggestion
					alt="" {/* This is a decorative image */}
```

jboland88 replied:
Hmmm, doing this:
```
alt="" {/* This is a decorative image */}
```
produces a syntax error for me.

I think escaping comments with `{}` is needed when placing them as a child of a JSX element, but not in the props declaration of an element 🤔 

Would be needed in this case:
```
<div>
    { /* A comment about the text */ }
    <p>Some text</p>
<div>
```

Does that seem correct?

sirbrillig replied:
Yeah, it's a topic I'm not certain about, but I think your suggestion would work and would feel better to me.

jboland88 replied:
Done.

sirbrillig commented on client/components/marketing-survey/cancel-jetpack-form/index.tsx:
```diff
@@ -193,11 +206,16 @@ const CancelJetpackForm: React.FC< Props > = ( {
+
+		// When an offer has been accepted, redirect back to the purchases page.
+		if ( cancellationOfferApplySuccess || steps.OFFER_ACCEPTED_STEP === cancellationStep ) {
+			page.redirect( props.purchaseListUrl );
```
Do you really want to redirect using `page.redirect()` and not just `page()`? IIRC, `page.redirect()` removes the current page from the browser history. I'm just asking because I've seen it used accidentally before.

sirbrillig commented on client/components/marketing-survey/cancel-jetpack-form/jetpack-cancellation-offer-accepted.tsx:
```diff
@@ -0,0 +1,38 @@
+import { useTranslate } from 'i18n-calypso';
+import * as React from 'react';
```
Same as above, this could just import the type.
```suggestion
import type { FC } from 'react';
```

jboland88 replied:
Nice one!

sirbrillig commented on client/state/cancellation-offers/selectors/get-cancellation-offer-apply-error.js:
```diff
@@ -0,0 +1,10 @@
+ *
+ * @param state
+ * @param purchaseId
+ * @returns {any}
```
This is not TS but even so we should try to avoid using the `any` type. Do you not know what this returns? From the code, I think its return type might be: `null | { code: string }`.

jboland88 replied:
Yes, that return type should be correct. Will update 👍

sirbrillig commented on client/state/cancellation-offers/selectors/get-cancellation-offer-apply-success.js:
```diff
@@ -0,0 +1,10 @@
+ *
+ * @param state
+ * @param purchaseId
+ * @returns {any}
```
Similar to above, I think we actually know this return type.
```suggestion
 * @returns {boolean}
```

jboland88 replied:
thanks! will update 👍

jboland88 committed:
Merge branch 'add/offer-step-in-cancellation-flow' into add/routine-to-claim-cancellation-offer

jboland88 committed:
Updates from feedback

jboland88 committed:
Move inline comment

jboland88 commented:
Thanks for the feedback @monsieur-z !

UI Feedback:
- Heading could benefit from preventWidows, and maybe from having a max-width.
I added a max width here, I originally could not figure out how to implement `preventWidows`, but I think I see it now. I can add that as a follow-up item.

- Could we maybe make the new discount and prices bold, so that they stand out more?
Sure thing 👍 I've updated this in the base PR and merged it in to this one.

- The links how your subscription works and how to cancel open a new modal, which forces the user to restart the cancellation process from scratch. Could we link to the associated articles directly instead?
Sure, updated this in the base PR and merged in since that's where those links were added. 

On the purchase screen, you mentioned:
> This may just be my brain: it took me a little while to understand that the highlighted sentence below was actually 2 separate pieces of information. Could we reformulate it, or at least separate it in 2 lines?

🤔 Good idea, let's tackle that in a separate PR.

jboland88 committed:
Disable cancellation button when offer is applying, dont retry when fetching offers

monsieur-z reviewed (approved):
Changes look good! It's ready to ship as far as I'm concerned. :shipit:
